### ISSUE TITLE
OBB file on extSD

### ISSUE THREAD

**Issue by BobAchgill** [agamemnus/cordova-plugin-xapkreader#122]
Not sure if this is an issue but I don't seem to be able to manually test my OBB file when it is on the external SD.  Works OK when on the internal SD.  Should it work on the External SD using these or other settings?  I would like to target the extSD ... just because there is more space there!

I am using ...
Command line : 

      cordova platform add android@6.0.0

config.xml  :

    <preference name="android-maxSdkVersion" value="24" />
    <preference name="android-installLocation" value="preferExternal" />

**Comment by agamemnus** [comment]
Not sure. It should work theoretically. Is that the settings for Cordova 6.5? Try it on that version first...

**Comment by BobAchgill** [comment]
I am using Cordova 6.5.0

Surely someone has experience using the xapkreader plugin and stores the obb files on the extSD.

I tried looking in stack overflow with this search...
Cordova expansion file extsd

... but no hits.

**Comment by agamemnus** [comment]
Yeah I think someone else wrote about this in the issues pages? It just gets really complex. I completely forgot all the details.

**Comment by BobAchgill** [comment]
Does xapkreader do anything in the code to support accessing the OBB file if the it is not located on the internal SD but instead located on the extSD?  Or does Android OS just handle switching the path so that the app is agnostic as to on which storage the the OBB is stored?

**Comment by agamemnus** [comment]
There is indeed code there to figure out the path...

**Comment by BobAchgill** [comment]
... but is there the switch code between internal and extSD?

Who originally wrote the xapkreader plugin?  Are they still around?

**Comment by agamemnus** [comment]
Intel. But then I reworked it.

https://github.com/agamemnus/cordova-plugin-xapkreader/blob/cordova-6.5.0/src/android/XAPKExpansionSupport.java

That file has the code for figuring out the path...

**Comment by BobAchgill** [comment]
OK thanks!
I will try to see if this solves the problem of making the cordova app with OBB file movable to the SD card....
https://ourcodeworld.com/articles/read/171/how-to-enable-the-move-to-sd-card-feature-in-a-cordova-app-for-android

**Comment by BobAchgill** [comment]
The link above had me add the cordova-custom-config plugin which made sure the android:installLocation="preferExternal" is automatically taken from the Config.xml and "properly" added to the manifest so the app can be movable by the user to and from the External/Internal storage.

So now when i look at the app's storage settings [on my S7]  I get the option to move the app to the external storage.  It dutifully runs a gas gage showing that the app (and data?) are being moved to the external storage.  When complete the app storage location indicator changes from Internal to External.

I looked at the external storage with a file manager and no folder was made for the app data being moved to external.  When I run the app it [runs!] but appears to still be using the internal obb file.  I know this because when I rename the internal obb app folder the links to the obb file no longer work.

Any ideas why the obb folder/file would not have been copied over?

When I try manually copying the obb folder/file over to the correct location on the external SD the app does not try to read it there.

I thought maybe it is a permissions thing... but the manifest shows this line...
`<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />`

Is there a forum where would be best to ask for help on this?

**Comment by agamemnus** [comment]
Android Central, perhaps.

**Comment by BobAchgill** [comment]
I posted this to Android Central...
https://forums.androidcentral.com/android-apps/883693-expansion-file-not-moved-extsd.html#post6226447

> Expansion file not moved to extSD
> 
> When I try to use the app storage setting menu on my S7 to move my app made with Cordova from internal to external storage the OBB expansion file does not move over though the app storage settings indicator does change from internal to external.

**Comment by agamemnus** [comment]
I just don't know the appropriate settings here, and it changes all the
time to boot...

**Comment by BobAchgill** [comment]
Also posted the question on the Android Forums...
https://androidforums.com/threads/cordova-obb-expansion-file-not-moved-to-external-storage.1264878/

**Comment by BobAchgill** [comment]
Looks like, from Android5.0 onwards,  cordova-plugin-xapkreader plugin would need to be upgraded to support using the serial number of the extSD in the path in order to support moving the app and OBB to the extSD...

> [From Android 5.0 onwards, the location of the external (removable) SD is no longer a fixed path. Instead, the serial number of the SD card is used in the path. For example, on my Samsung Galaxy S4 which is running Android 7.1.1, the physical external removable SD card path is /storage/4975-1401/.](https://stackoverflow.com/questions/45770040/external-storage-path-sd-card-for-android-5-1-1-and-later-with-cordova?rq=1)

... or is there a way to hack so that cordova-diagnostic-plugin can be used to make cordova-plugin-xapkreader redirect to the correct place to both move/write and read the externally stored OBB?

**Comment by BobAchgill** [comment]
I posted this feature request to cordova-diagnostic-plugin for some how it to help cordova-plugin-apkreader in storing app and OBB expansion files on external storage.
[https://github.com/dpa99c/cordova-diagnostic-plugin/issues/299](https://github.com/dpa99c/cordova-diagnostic-plugin/issues/299)

**Comment by BobAchgill** [comment]
Dave Alden with cordova-diagnostic-plugin says:

> getExternalSdCardDetails() will provide the filepath(s) to external removable storage locations.
> 
> These filepaths can be used along with cordova-plugin-file to read/write files to this location. Or you could enhance cordova-plugin-xapkreader to do this.
> 

getExternalSdCardDetails() is a function in his cordova-diagnostic-plugin.

@agamemnus are you able to add external (removable) storage support to cordova-plugin-xapkreader so that apps (Android 5 and beyond) can have their install moved to extSD ?

**Comment by agamemnus** [comment]
Maybe you can try updating the code and then submit a PR? I am very busy... if not, maybe the weekend.

**Comment by BobAchgill** [comment]
I'm not a developer. :(

Most of my users in developing countries will not have enough internal free space to install the free  1.5GB literacy app.  Hence, my quest to get the support for ...

`<preference name="android-installLocation" value="preferExternal" />`

to work.

One other random thought... Can I just compile using  something less than android@5 and skirt the issue that way?  Or does this issue result because the device's Android level is beyond 5?

**Comment by agamemnus** [comment]
The latter, I am pretty sure.

**Comment by BobAchgill** [comment]
Your donor gittip link does not work in your readme.

**Comment by agamemnus** [comment]
Ha. Yeah they shut the system down. How about I will update that when I
also update the file system code?

**Comment by var18k** [comment]
Yeah same problem here. Pls give a solution for how to move obb on SD.

**Comment by BobAchgill** [comment]
@agamemnus any chance you can implement this fix soon?

**Comment by agamemnus** [comment]
`getExternalSdCardDetails` is a function on cordova-diagnostic-plugin... so I would have to copy their code.... not an MIT license..... bit of a headache...... might have to do my own.

Maybe this weekend, sorry.

**Comment by BobAchgill** [comment]
Any progress over the weekend?

**Comment by agamemnus** [comment]
I am on a deadline for a thing, so no progress.

**Comment by BobAchgill** [comment]
Maybe this weekend?

**Comment by agamemnus** [comment]
Maybe.

**Comment by BobAchgill** [comment]
Maybe this week?

**Comment by agamemnus** [comment]
Maybe...

**Comment by agamemnus** [comment]
Bob,

Thank you so much for waiting. I hope you still need this. I have earmarked some time this weekend to work on it and help you solve this issue.

**Comment by BobAchgill** [comment]
Wonderful!
Yes, I still need this.

**Comment by BobAchgill** [comment]
Any progress?!

**Comment by agamemnus** [comment]
Unfortunately I had an unexpectedly large amount of other work on the
weekend. I'm about to go to sleep. Maybe, maybe, tomorrow...

**Comment by agamemnus** [comment]
Looks like I had a pull request open with some changes to the code that might fix it. Does it work?

**Comment by BobAchgill** [comment]
Tell me how to use that version as opposed to the normal version and I'll test it!

**Comment by agamemnus** [comment]
It's the same now (i.e.: 6.5).

**Comment by BobAchgill** [comment]
There is progress...
I tested the pull on an Android 7 device...good news/bad news...
1) The obb dir/file does NOT get moved when app is moved to extSD using move control from the app Storage location button.
2) However, now the app WILL read the OBB file on the extSD if placed there manually!
3) Interesting that the app DOES delete the OBB file on the extSD when the app is deleted.


For some reason the expansion file does not work at all... either internal or extSD when I tested on an Android 4.4.2 device.

**Comment by agamemnus** [comment]
So then the question is, what actually happens when you press that button? Maybe there is a function that runs that we could implement to copy over the files?

The 4.4.2 issue is troubling, also.... but presumably, only a tiny fraction of users (er, 10.5% as of April...) are still on version 4 of Android, as version 5 was made public 3 years ago....

**Comment by BobAchgill** [comment]
Correction on my testing...
4.4.2 DOES work but it persists in only using the internal OBB expansion file even when it should be using the external OBB expansion file. 

The adding code to move the OBB directory and expansion file to the extSD and back sounds good.  So the Android OS doesn't do that??

I say "move " because that frees up the space on the internal memory... the whole reason for moving an app to the extSD.  And being able to move back... because the user sees a button that allows that round trip.

**Comment by agamemnus** [comment]
Maybe not: http://www.tomshardware.com/answers/id-2798563/move-obb-data-files-andriod-card.html

Perhaps one way is to check the obb folder on startup... if it doesn't have the files, check the other location, see if they are there, and then (if they're there) ask the user if they want to move the files. If they aren't, then re-download the files...

But at this point, isn't the primary problem solved?

**Comment by BobAchgill** [comment]
Original Question...
"Not sure if this is an issue but I don't seem to be able to manually test my OBB file when it is on the external SD."

I don't know the stats for Android 4 devices in developing countries.  That's my target.  It would be nice if "they" were included in the "fix" if it's easy to determine why the code on Android 4 devices is still looking for the OBB expansion file on the internal storage when the app has been moved to extSD.  The cheaper phones have less internal storage hence the reason I'm trying to give the user the extSD app/data store option.  

If you want me to open a separate issue for the remaining questions ... I can.

Remaining parts...
1) move OBB expansion dir/file to/from extSD when app moved
2) access OBB expansion file on extSD for Android 4.x

Though these two issues seem to be integral to the subject issue.

I see that your idea for moving the data will work but why not use the app move button to trigger the move process?  Does the move button force a shutdown of the app as its moved?  If so then yes... that solution will be good.  Otherwise, wouldn't the user be left wondering why the data did not get moved when the app was moved?

And one more
3) support app install preference to target OBB storage on internal vs extSD.

**Comment by agamemnus** [comment]
"I see that your idea for moving the data will work but why not use the app move button to trigger the move process?"

Well, that would only work if there was some instant way for the app to know that happened. There might or might not be a way. I couldn't find a way yet. But definitely on startup it would know if the files were missing...

I am still lost on how this process that you want should happen and what happens now... I think you mentioned it before, but I forgot. So the user downloads it from the app store and then you tell them to move it? Or...

**Comment by BobAchgill** [comment]
This will be added to the config so that the install will default prefer installing the apk and OBB externally if the extSD exists.  The plugin will find this install setting in the manifest.  If the app developer chooses not to set this I think it defaults to "auto" or "internal".


<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    android:installLocation="preferExternal"
    ... >

Later the user can opt to move the app/OBB expansion data to internal using the application storage option button.


Reference:
https://developer.android.com/guide/topics/data/install-location

**Comment by agamemnus** [comment]
But does the user even know this option exists though? What % of the time are they trying to do it?

If this is something a lot of users want, maybe it would be better to ask directly in the app (on launch) where the user wants it stored.

**Comment by BobAchgill** [comment]
It seems to be pretty much a standard that apps that are extSD storage enabled give the control button in the app storage setting window.  It was hard for me to find that button but that is where it is... typically.

As far as offering the choice of where to store the expansion file (internal or extSD) at initial launch...
What you suggest would be better than just defaulting them to whatever the app developer wants to set as default.
Are you able to check how big the expansion file is before download so that you can give them stats on where it might fit the best?

If you can't give the user storage footprint comparison stats then I'm fine with going with using the Storage preferred location setting in the config.

95 percent of my users will never have enough internal room to store the 1.5GB expansion file... so asking them at launch with stats will just be a good reminder that they will soon need larger extSD. :)

**Comment by agamemnus** [comment]
Normally, the obb download happens immediately after clicking to download the file from Google Play (at the same time as the apk) ...... I'm not sure what's going on anymore. I forgot if you were even getting the file from Google Play... presumably, you are...

**Comment by BobAchgill** [comment]
Correct.  "The obb download happens immediately after clicking to download the file from Google Play (at the same time as the apk)."

What I am recommending is that the plugin support the preferred storage location attribute that the developer sets in the config.  Maybe it does already do this??

I don't know because I have been doing manual tests (without Google play).

**Comment by BobAchgill** [comment]
Remaining issues...
1) support storage location move button
2) Android 4.x
3) *support preferred storage location attribute in manifest on initial download.

* maybe it already does this

**Comment by agamemnus** [comment]
I'm just saying, if there is no space it I think will tell you and won't
download it, I think. Maybe it will ask if you want to install it in the
external drive space. Unclear, memory foggy. You should really just publish
it in some way (maybe as a beta test) to test it out.

**Comment by BobAchgill** [comment]
Ok!
How about 1 an 2?

**Comment by agamemnus** [comment]
Ok, can try but it just doesn't seem too important if the users will use Google Play as they probably won't be able to download it without sufficient space anyway...

So it wasn't working on 4.x before? Or was? (Are you only testing with 4.x?)

**Comment by agamemnus** [comment]
(Although you can change settings on Google Play to set preferred install location... and it just might go to the area with the available space, so seems like a non-issue except for the 4.x problem...)

**Comment by BobAchgill** [comment]
I did not test with 4.x before.

**Comment by BobAchgill** [comment]
Looks like the method for setting Google Play to download to ExtSD is a hack.  It takes over the ExtSD and wipes all other data.  Does not appear to be a Google recommended method.  And may not work on all devices.

http://www.tomsguide.com/forum/id-3439437/set-downloading-location-play-store-apps-directly-card-samsung-galaxy-grand-prime.html

**Comment by BobAchgill** [comment]
I'll do a full system test with downloading from Google play.  My guess is that the plugin code handles the download of the OBB file and writes to storage.  .... if so then you would make the same changes you made to get Android 5+ to work for reads of ExtSD data.... to the downloader for writes to the extsd.

**Comment by BobAchgill** [comment]
So far I have tested download from Google play using a Android 4.2.2... here is what happened...
I compiled using the default storage location being set to prefer extSD.
The download apparently put the app on the extSD but the OBB was written to the internal storage.
I say that the app was written to the extSD because when I look at the app storage settings it shows that the app is stored on the extSD.
When I try to run the freshly downloaded app it immediatly shows a dark screen with the message...
"Downloading assets..." but then shows "Unfortuanatley, Test has stopped."

Next, 
I moved the OBB/file over to the extSD using file manager and tried running the app again to see if putting the data in the "expected" location on the extSD would prevent the app from bombing trying to download ... it still failed with the same error message.

Next, 
I pressed the move button in the app settings storage to move the app over to the internal storage.  The move occurs.  At least the gas gage counts down the move and concludes with showing that the app is now located on internal storage.  But when I look with file manager the OBB/file is not moved over to internal storage.  

So Next, I move the OBB/file over with file manager to the expected location in internal storage... And when I try to run the app again it still fails trying to download assets.

It appears that either way the app can not read the obb/file on either internal or extSD storage so it trys to download the file??

I will try the Android 7.0 test tomorrow.

**Comment by BobAchgill** [comment]
Test on Android 7 fails same way...
"Downloading assets..." but then shows "Unfortuanatley, Test has stopped."

**Comment by agamemnus** [comment]
Something isn't right .... should be working since the latest patch came from a user for whom it works. Can you make a link to the file?

**Comment by BobAchgill** [comment]
https://play.google.com/apps/internaltest/4699174589605622933

I may need to approve your email address? Give it a try.

**Comment by agamemnus** [comment]
Ok [redacted].

**Comment by BobAchgill** [comment]
Added

**Comment by agamemnus** [comment]
OK. Doesn't work for me either. Will investigate tomorrow.

**Comment by agamemnus** [comment]
I have seen people mention this kind of thing before... any way you could send me the uncompiled file to look at so I could check the settings?

**Comment by BobAchgill** [comment]
It appears to work with Android 7 now.
With one caveat...
When I move the app to external storage ... the app moves and the obb copy (i.e. does not free up internal storage space) over to ext storage but it still looks for the OBB on the internal storage... and if I delete the OBB from ext storage it starts downloading the expansion file again from Google play.

The download of the app does not appear to honor the storage location pointer to have the app prefer to install on external storage.

This makes it very difficult if not impossible for the user who does not have enough internal storage to download the app.

**Comment by agamemnus** [comment]
second part is expected behavior...... does it download to external storage?

**Comment by BobAchgill** [comment]
Testing with Android 7.x...

It will not install the app to the external storage using this...
`<preference name="android-manifest/@android:installLocation" value="preferExternal" />`

Once it is installed it is possible to move the app to external storage but not the obb.

So there is no scenario where the obb can be stored/accessed on the ext storage.

My guess is that it did not initially install the app to the preffered ext storage because it knew that it could not also write to the obb file to ext storage??

**Comment by agamemnus** [comment]
Huh. Ok, I'll do some more tests tomorrow.

**Comment by BobAchgill** [comment]
With preferred install storage location set to extSD...
`<preference name="android-manifest/@android:installLocation" value="preferExternal" />`

Test case 1: Android 7.x 
App installed to internal And OBB to internal.
Can move app to extSD but not OBB 

Test case 2: Android 4.4.2 
App installed to extSD but OBB to internal.

What is hoped for is that some how the OBB can be stored and accessed on the extSD on initial install to save space on the User's internal storage.
